pkg/errors/stacktrace.go

```diff
@@ -0,0 +1,48 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package errors
+
+import (
+	"fmt"
+	"runtime"
+	"strings"
+)
+
+// stacktrace holds a snapshot of program counters.
+type stacktrace []uintptr
+
+// newStackTrace captures a stack trace. It skips first 3 frames to record the
+// snapshot of the stack trace at the origin of a particular error. It tries to
+// record maximum 16 frames (if available).
+func newStackTrace() stacktrace {
+	const stackDepth = 16 // record maximum 16 frames (if available)
+
+	pc := make([]uintptr, stackDepth)
```

Comment:
This implementation has the same issue as `pkg/errors` i.e. the `pc` escapes to heap thus the entirety of the 16 uintptr is kept in memory even if only the stack is only 1 call frame deep.  I have explained the issue in details [here](https://github.com/emperror/errors/issues/27) based on a real-world scenario. 



Please try `go build -gcflags='-m -m' ./pkg/errors/stacktrace.go` and you would find 

```
pkg/errors/stacktrace.go:21:12: make([]uintptr, stackDepth) escapes to heap:
pkg/errors/stacktrace.go:21:12:   flow: pc = &{storage for make([]uintptr, stackDepth)}:
pkg/errors/stacktrace.go:21:12:     from make([]uintptr, stackDepth) (spill) at pkg/errors/stacktrace.go:21:12
pkg/errors/stacktrace.go:21:12:     from pc := make([]uintptr, stackDepth) (assign) at 
```
The solution is to allocate the buffer `pc` on stack like the original implementation and then copy whats needed off it to prevent escaping to heap. 

```suggestion
	const maxDepth = 16
	var pcs [maxDepth]uintptr.  // allocate on stack

	n := runtime.Callers(3, pcs[:])
	st := make(stack, n)
	copy(st, pcs[:n])
	return st
```

Another thought is if we really need to hold on to the entire CallFrame until `String()` is called? At the expense of some compute, we could calculate Stacktrace string immediately.

Reply:
Nice nitpick and great catch @sthaha.
>  the pc escapes to heap thus the entirety of the 16 uintptr is kept in memory even if only the stack is only 1 call frame deep.

Yes, you are right - there is no point in holding an extra chunk of unused heap memory, And I think you are copying to another temporary slice because even though we are returning `stacktrace(pc[:n])` the capacity of PC still remains 16 (allocated memory), right?

But this additional copy adds some extra complexity : ) (though it's very minor and negligible as we are dealing with length of 16) From escape analysis the function complexity gets increased from 88 -> 96
```
./stacktrace.go:20:6: cannot inline newStackTrace: function too complex: cost 96 exceeds budget 80
.
.
./stacktrace.go:20:6: cannot inline newStackTrace: function too complex: cost 88 exceeds budget 80

```

Since go1.2 we have 3 index slice capability where the third param can define the capacity of the newly created slice.

So just updating the return statement to `return stacktrace(pc[:n:n])` would yield the same effect of the suggestion you proposed.

Thanks a lot. That was really awesome.

> Another thought is if we really need to hold on to the entire CallFrame until String() is called? At the expense of some compute, we could calculate Stacktrace string immediately.

I think we shouldn't do it for the following reasons
- Its compute expensive - runtime needs to retrieve caller information
- More memory usage - the `String` method is meant for human consumption so naturally, this adds a lot of extra information, text etc compared to the 16 element slice of type unitptr. During error changing it would be worse.
- Not every error needs the stacktrace. For specific format verbs like `%+v` stacktrace gets dumped recursively (only then String gets called).

Reply:
> Since go1.2 we have 3 index slice capability where the third param can define the capacity of the newly created slice.
💯  .. Thank you, I wasn't aware :) 

> I think we shouldn't do it for the following reasons ... 

> Its compute expensive - runtime needs to retrieve caller information 
Have you tried a benchmark? 

> More memory usage - the String method is meant for human consumption so naturally, this adds a lot of extra information, text etc compared to the 16 element slice of type unitptr. During error changing it would be worse.

What I meant was to only hold on to information that you need than have `uintptr` which now points to the callframe and IIUC, holding onto the callframe  will now prevent GC from cleaning the callframe info until the error is garbage collected as well. 

It may be worth seeing the inuse allocation if we don't hold on to the callframes.

Reply:
> > I think we shouldn't do it for the following reasons ...
> 
> > Its compute expensive - runtime needs to retrieve caller information
> 
> Have you tried a benchmark?
> 

IIUC, in the new approach that you are suggesting we should call the `String` immediately at the expense of an extra call which might not be required if the error is not logged with an `"%+v"` format verb. So we are adhering to extra computation which might not be of any use. So obviously it will be expensive in time against the current approach  as the current one only calls string when it's required.

> > More memory usage - the String method is meant for human consumption so naturally, this adds a lot of extra information, text etc compared to the 16 element slice of type unitptr. During error chaining it would be worse.
> 
> What I meant was to only hold on to information that you need than have `uintptr` which now points to the callframe and IIUC, holding onto the callframe will now prevent GC from cleaning the callframe info until the error is garbage collected as well.
> 

Now coming to the memory footprint, IMHO storing unstructured data (the stack trace output string) is generally a bad idea. If we change the output string format from `fmt.Sprintf("> %s\t%s:%d\n", frame.Func.Name(), frame.File, frame.Line)` this to something else memory footprint gets changed. So I think this optimization won't be valid in longer run.

After giving it a thought, I think it's a tradeoff, why?
Memory usage:
- current approach: 16*4 bytes uintptr + unreleased memory from gc (not sure what it stores on that PC address, might be some function related metadata to populate the callFrames. definitely releases ass the resources used inside the function)
- proposed approach: large unstructured string + still some unreleased memory (not sure when will the next cycle of go GC will kick off and it's dependent on the config). 

I am not sure if we can benchmark this anyhow or not.

>the error is garbage collected as well.

If you see, in Go world err has the comparatively shortest lifespan than anything else. In a well written program they gets handled immediately by being returned to the caller function or logged into the sink. So I think we are good here. After all it's an internal package and if we found some untoward runtime behaviour we can always do different optimization. 

Thank you : )

Reply:
> If you see, in Go world err has the comparatively shortest lifespan than anything else. 

Not always. Check our fixed issue: https://github.com/thanos-io/thanos/issues/5257

Anyway, I asked @sthaha to remind us about the optimization emperor did - to make sure we are aware. As usually we can iterate over it, we won't do it perfectly over time - the main thing is to get APIs as best as possible - it's hard to change them later.